core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/SplitsMapping.java

```diff
@@ -0,0 +1,299 @@
+                Map<Integer, ImmutableList.Builder<Split>> updatesSplitsMap = updates.get(planNodeId);
+                for (Map.Entry<Integer, ImmutableList.Builder<Split>> innerEntry : updatesSplitsMap.entrySet()) {
+                    Integer sourcePartitionId = innerEntry.getKey();
+                    List<Split> sourceSplits = originalSplitMap.get(sourcePartitionId);
+                    if (sourceSplits != null) {
+                        targetSplitsMapBuilder.put(sourcePartitionId,
+                                ImmutableList.<Split>builder().addAll(sourceSplits).addAll(innerEntry.getValue().build()).build());
+                    }
+                    else {
+                        // no source splits for give sourcePartitionId
+                        targetSplitsMapBuilder.put(sourcePartitionId, innerEntry.getValue().build());
+                    }
+                }
+
+                for (Map.Entry<Integer, List<Split>> innerEntry : originalSplitMap.entrySet()) {
+                    Integer sourcePartitionKey = innerEntry.getKey();
+                    if (updatesSplitsMap.containsKey(sourcePartitionKey)) {
+                        // already handled above
+                        continue;
+                    }
+                    // reuse source immutable split list
+                    targetSplitsMapBuilder.put(sourcePartitionKey, innerEntry.getValue());
+                }
+                immutableSplitsBuilder.put(planNodeId, targetSplitsMapBuilder.build());
+            }
+
+            for (Map.Entry<PlanNodeId, Map<Integer, ImmutableList.Builder<Split>>> entry : updates.entrySet()) {
+                PlanNodeId planNodeId = entry.getKey();
+                if (originalMapping.splits.containsKey(planNodeId)) {
+                    // handled above
+                    continue;
+                }
+
+                ImmutableMap.Builder<Integer, List<Split>> targetSplitsMapBuilder = ImmutableMap.builder();
+                for (Map.Entry<Integer, ImmutableList.Builder<Split>> innerEntry : entry.getValue().entrySet()) {
+                    targetSplitsMapBuilder.put(innerEntry.getKey(), innerEntry.getValue().build());
+                }
+                immutableSplitsBuilder.put(planNodeId, targetSplitsMapBuilder.build());
+            }
+
+            return new SplitsMapping(immutableSplitsBuilder.buildOrThrow());
+        }
+    }
+
+    private static class NewBuilder
+            implements Builder
+    {
+        private final Map<PlanNodeId, Map<Integer, ImmutableList.Builder<Split>>> splitsBuilder = new HashMap<>();
+
+        @Override
+        public Builder addSplits(PlanNodeId planNodeId, int partitionId, List<Split> splits)
+        {
+            splitsBuilder.computeIfAbsent(planNodeId, ignored -> new HashMap<>())
+                    .computeIfAbsent(partitionId, ignored -> ImmutableList.builder())
+                    .addAll(splits);
```

Comment:
Did you consider `Table`?

Reply:
Like
```
        Table<PlanNodeId, Integer, ImmutableList.Builder<Split>> ....;
```

It would work - but Table interface is less nice to work with. E.g you do not have `computeIfAbsent`